libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,447 @@
+
+#ifndef PLONK_PPZKSNARK_TCC_
+#define PLONK_PPZKSNARK_TCC_
+
+namespace libsnark
+{
+
+template<typename FieldT> void print_vector(std::vector<FieldT> v)
+{
+    for (size_t i = 0; i < v.size(); ++i) {
+        printf("[%2d]: ", (int)i);
+        v[i].print();
+    }
+}
+
+//
+// Compute the Lagrange basis polynomails for interpolating sets of
+// n points
+//
+// INPUT:
+//
+// - npoints - number of points
+//
+// OUTPUT:
+//
+// - L[0..n-1][0..n-1]: Lagrange basis over the n roots of unity
+//                      omega_0, ..., omega_{n-1} i.e. L[omega_i] =
+//                      [a0, a1, ..., a_{n-1}] is a vector
+//                      representing the coefficients of the i-th
+//                      Lagrange polynomial L_i(x) =
+//                      a0+a1x+a2x^2+..+a_{n-1}x^{n-1}
+//                      s.t. L_i(x=omega_i)=1 and
+//                      L_i(x\neq{omega_i)})=0
+//
+// Note: uses libfqfft iFFT for the interpolation
+//
+template<typename FieldT>
+void plonk_compute_lagrange_basis(
+    size_t npoints, std::vector<polynomial<FieldT>> &L)
+{
+    assert(L.size() != 0);
+    assert(L.size() == L[0].size());
+    assert(L.size() == npoints);
+
+    std::shared_ptr<libfqfft::evaluation_domain<FieldT>> domain =
+        libfqfft::get_evaluation_domain<FieldT>(npoints);
+    for (size_t i = 0; i < npoints; ++i) {
+        polynomial<FieldT> u(npoints, FieldT(0));
+        u[i] = FieldT(1);
+        // compute i-th Lagrange basis vector via inverse FFT
+        domain->iFFT(u);
+        L[i] = u;
+    }
+}
+
+//
+// Interpolate a set of points as a polynomial over Lagrange basis
+//
+// INPUT:
+//
+// - f_points[0..n-1]: a set of points (0,y0), (1,y1),
+//                     ... (n-1,y_{n-1}) s.t. y0=f_points[0],
+//                     y1=f_points[1], ... which we want to interpolate
+//                     as a polynomial
+//
+// - L[0..n-1][0..n-1]: Lagrange basis over the n roots of unity
+//                      omega_0, ..., omega_{n-1} i.e. L[omega_i] =
+//                      [a0, a1, ..., a_{n-1}] is a vector
+//                      representing the coefficients of the i-th
+//                      Lagrange polynomial L_i(x) =
+//                      a0+a1x+a2x^2+..+a_{n-1}x^{n-1}
+//                      s.t. L_i(x=omega_i)=1 and
+//                      L_i(x\neq{omega_i)})=0
+//
+// OUTPUT:
+//
+// - f_poly[0..n-1]: the coefficients [a0, a1, ..., a_{n-1}] of the
+//                   polynomial f(x) interpolating the set of points
+//                   f_points over the Lagrange basis. For example if
+//                   f_poly[0..n-1] = [a0, a1, ..., a_{n-1}] then
+//                   this represents the polynomial f(x) =
+//                   \sum^{n-1}_{i=0} f_vec[i] * L[i] =
+//                   a0+a1x+a1x^2+...+a_{n-1}x^{n-1} such that
+//                   f(omega_i)=f_vec[i].
+//
+template<typename FieldT>
+void plonk_interpolate_over_lagrange_basis(
```

Comment:
This function signature seems to accept vectors passed by value (potentially incurring many copies), and accepts the "return" value as the second argument.  Let's pass vectors as const refs, and either return the output value, or keep "output" arguments consistently as the last param.

Reply:
Fixed in all functions, in all files.